google/auth/credentials.py

```diff
@@ -0,0 +1,209 @@
+    construction. Some classes will provide mechanisms to copy the credentials
+    with modifications such as :meth:`ScopedCredentials.with_scopes`.
+    """
+    def __init__(self):
+        self.token = None
+        """str: The bearer token that can be used in HTTP headers to make
+        authenticated requests."""
+        self.expiry = None
+        """Optional[datetime]: When the token expires and is no longer valid.
+        If this is None, the token is assumed to never expire."""
+
+    @property
+    def valid(self):
+        """Checks the validity of the credentials.
+
+        This is True if the credentials have a :attr:`token` and the token
+        is not :attr:`expired`.
+        """
+        return self.token is not None and not self.expired
+
+    @property
+    def expired(self):
+        """Checks if the credentials are expired.
+
+        Note that credentials can be invalid but not expired becaue Credentials
+        with :attr:`expiry` set to None is considered to never expire.
+        """
+        now = _helpers.utcnow()
+        if self.expiry is None or self.expiry > now:
+            return False
+        else:
+            return True
+
+    @abc.abstractmethod
+    def refresh(self, request):
+        """Refreshes the access token.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+
+        Raises:
+            google.auth.exceptions.RefreshError: If the credentials could
+                not be refreshed.
+        """
+        # pylint: disable=missing-raises-doc
+        # (pylint doesn't recognize that this is abstract)
+        raise NotImplementedError('Refresh must be implemented')
+
+    def apply(self, headers, token=None):
+        """Apply the token to the authentication header.
+
+        Args:
+            headers (Mapping): The HTTP request headers.
+            token (Optional[str]): If specified, overrides the current access
+                token.
+        """
+        headers[b'authorization'] = 'Bearer {}'.format(
+            _helpers.from_bytes(token or self.token))
+
+    def before_request(self, request, method, url, headers):
+        """Performs credential-specific before request logic.
+
+        Refreshes the credentials if necessary, then calls :meth:`apply` to
+        apply the token to the authentication header.
+
+        Args:
+            request (google.auth.transport.Request): A callable used to make
+                HTTP requests.
+            method (str): The request's HTTP method.
+            url (str): The request's URI.
+            headers (Mapping): The request's headers.
+        """
+        # pylint: disable=unused-argument
+        # (Subclasses may use these arguments to ascertain information about
+        # the http request.)
+        if not self.valid:
+            self.refresh(request)
+        self.apply(headers)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ScopedCredentials(object):
```

Comment:
Ah! I think I finally I see what you're after, you're essentially wanting to hide the concrete classes altogether and only expose the public interfaces, so we'd end up with something like this:

```python

class PublicInterface(object):
    pass

# may or may not be needed, but if it is, this is where it'll be.
class _PrivateCommonConcreteStuff(PublicInterface):
   pass

class _PrivateConcrete(_PrivateCommonConcreteStuff):
   pass

def constructor_for_thing():
    return _PrivateConcrete()

```


This sort of implementation hiding seems way over the top and unpythonic to me. Granted, I'm not (nor have ever claimed to be) an expert in object-oriented design, but it seems really unintuitive as a user of the library. It's also somewhat unintuitive as the maintainer of the library (I mean, it took you no less than 3 very long comments to express this to me, and I'm still unsure if I'm 100% clear on it).

I'm happy to meet in person to discuss, or if you feel really confident that there's no way I'll talk you out of this then I'll relent and implement it as above. I'd also like to hear @dhermes's thoughts on it.

Some specific responses:

> Including a public class in an API is a big commitment - much bigger than including an interface describing what the API user may do with a type and one or more functions available for constructing objects that implement the type. 

This seems like a viewpoint coming from a Java-centric world. Not saying it's wrong or that it doesn't have advantages (it does, and I see that now), but it does feel unfamiliar in Python.

>  An alternative is to define extra interfaces that define no methods of their own but by inheriting from two or more others serve to specify a union of them

I get what you're saying here, but in terms of actual usage the consumers of this library aren't going to ask 'Is this a ComboCredential'? They're going to say 'Can this credential sign bytes' - so the combined interfaces would just be for the benefit of the docstrings. Perhaps it'll be better if I renamed `ScopedCredentials` and `SigningCredentials` to `HasScopes` and `CanSign` (I was thinking of doing so anyway).